fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/plans/RecordQueryChooserPlanBase.java

```diff
@@ -0,0 +1,135 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.record.query.plan.plans;
+
+import com.apple.foundationdb.record.RecordCoreArgumentException;
+import com.apple.foundationdb.record.RecordMetaData;
+import com.apple.foundationdb.record.query.plan.AvailableFields;
+import com.apple.foundationdb.record.query.plan.temp.CorrelationIdentifier;
+import com.apple.foundationdb.record.query.plan.temp.Quantifier;
+import com.apple.foundationdb.record.query.predicates.Value;
+import com.google.common.base.Suppliers;
+import com.google.common.base.Verify;
+
+import javax.annotation.Nonnull;
+import java.util.List;
+import java.util.Set;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/**
+ * A {@link RecordQueryChooserPlanBase} is a plan that, at runtime, chooses to execute plans from multiple alternatives.
+ * Subclasses of this base class will select, at runtime, one or more plans to execute from the list of optional children
+ * held in this base class.
+ * The invariant enforced by this base class is that all child plans are similar in the sense that they provide results
+ * that are interchangeable: The same result set, in a compatible order (if one is specified).
+ * Selecting one plan over the other will not have an impact on the client receiving the results.
+ */
+public abstract class RecordQueryChooserPlanBase implements RecordQueryPlanWithChildren {
+
+    @Nonnull
+    protected final List<Quantifier.Physical> quantifiers;
+    private final boolean reverse;
+    @Nonnull
+    private final Supplier<List<? extends Value>> resultValuesSupplier;
+
+    protected RecordQueryChooserPlanBase(@Nonnull final List<Quantifier.Physical> quantifiers) {
+        Verify.verify(!quantifiers.isEmpty());
+        this.quantifiers = List.copyOf(quantifiers);
+        boolean firstReverse = quantifiers.get(0).getRangesOverPlan().isReverse();
+        if (!getChildStream().allMatch(child -> child.isReverse() == firstReverse)) {
+            throw new RecordCoreArgumentException("children of chooser plan do not all have same value for reverse field");
+        }
+        this.reverse = firstReverse;
+        // Since all child plans are assumed to return the same Values, pick the first one to flow the values back.
+        this.resultValuesSupplier = Suppliers.memoize(quantifiers.get(0)::getFlowedValues);
```

Comment:
This is dangerous. The planner should inject a `PickValuesValue` here. I think I did this for Norse but essentially you create values as results here that look like this
```
PickValuesValue(ordinalQuantifiedColumn, all the zeroeth columns),
PickValuesValue(ordinalQuantifiedColumn, all the first columns),
PickValuesValue(ordinalQuantifiedColumn, all the second columns),
...
PickValuesValue(ordinalQuantifiedColumn, all the nth columns)
```

`PickValuesValue()` has `n+1` inputs, i.e. an integer (a selector) and n options that can be picked from. At runtime you set the ordinal to the option you want and then `eval` the value which gives you the correct datum.

All of this is necessary since we don't ever want to disturb the derivation graph of a value. If you do it the way you do it, transformations would only ever _see_ the `0` variant and are oblivious to others which we want to avoid.

Reply:
Done. Created `ValuePickerValue` that stores multiple sub-values and delegates to the `selected` one.